### ISSUE TITLE
Collinearity mitigation on StackingClassifier and StackingCVClassifier

### ISSUE THREAD

**Issue by soumyadsanyal** [rasbt/mlxtend#185]
Issue:

Arises when you implement / execute stacking on a dataset `D` by scoring the dataset using the `predict_proba` or `predict` methods for each 'input' model, then concatenate the scores as e.g.: `D' = concatenate([model.predict_proba(D) for model in base_models], axis=1)`.

Problem: since each matrix `model.predict_proba(D)` (for each `model`) satisfies `sum(model.predict_proba(D), axis=1) == np.ones(..)`, the resulting collinearity in the matrix `D'` may affect convergence or parameter estimates esp. when the stacking metaestimator is a regression model.

Propose:

Add an option to score the dataset as follows:

`D_ = concatenate([model.predict_proba(D)[:, -1] for model in base_models], axis=1)`.

**Comment by rasbt** [comment]
Thanks, I would really appreciate a PR here. Do you have a good "parameter" name in mind for the Stacking(CV)Classifier `__init__` for this option?

**Comment by soumyadsanyal** [comment]
How does `collinearity_guard` sound? (Very open to suggestions!)

**Comment by rasbt** [comment]
This sounds good to me :). Would have called it 'reduce_dof' (reduce degrees of freedom) or so, but collinearity_guard sounds better imho :). 

Btw right now you are thinking about removing the last proba, right? What do you think about accepting the following arguments for `collinearity_guard`?

- `None` (does nothing, should maybe be the default?)
- `-1` (last column)
- a positive integer < n_features: column with the specified index is left out
- `'random'`: leave out a random column for each model (would maybe make an interesting (novel?) "regularization" approach; thinking of dropout here ;))

I am happy to help with the PR if that sounds like too much work! :)

**Comment by soumyadsanyal** [comment]
Yes, that was the initial plan, but those suggestions sound good! (They won't be too hard to do, I'll take care of them in the morning :) ).

**Comment by soumyadsanyal** [comment]
@rasbt quick thoughts / a question on the `random` option:

Currently, it looks like the `StackingClassifier` and `StackingCVClassifier` classes are somewhat decoupled from the dataset they will be called on, in the sense that the dataset is not a parameter to `__init__`. In principle then it's possible to `fit` on a dataset and `predict` or `predict_proba` on a different dataset, even though that should throw an exception in the case the dimensions change, etc. (If I'm misunderstanding please correct me!)

To implement the `random` option, we'd need to make sure that the _same_ column is dropped in each of the `fit`, `predict` and `predict_proba` methods, so the realization of which column to drop needs to be uniform across these methods. For that reason the choice of which column to drop should be abstracted into a class function, or maybe decided in `__init__`. In particular, the choice depends on the actual dataset (as we need to select from one of its columns), so it needs the dataset (or at least `dataset.shape[1]`) as a parameter.

Do we want to implement a guarantee that e.g. `fit` and `predict` are only ever called on slices of the same dataset (perhaps by requiring the dataset as a parameter to `__init__`; of course, that comes with its own pros and cons)? Or should I just implement this functionality assuming that is the case?

(If I'm not misunderstanding this, I might implement the other options and do another PR for the `random` option if substantial refactoring is needed.)

**Comment by rasbt** [comment]
> Currently, it looks like the StackingClassifier and StackingCVClassifier classes are somewhat decoupled from the dataset they will be called on, in the sense that the dataset is not a parameter to __init__.

Yeah, that's to ensure compatibility with scikit-learn's classes (e.g., GridSearch, Pipeline, etc) -- it's a requirement within the scikit-learn API.

> the fit, predict and predict_proba methods, so the realization of which column to drop needs to be uniform across these methods, and should not change unless the model(s) is(are) refitted. 

good point. I think a good way to do that would be do determine this random column in the `fit` method E.g., 

    fit(self, X, y):
        ...
        rng = np.random.RandomState(self.random_state)
        self.random_column_index = rng.randint(X.shape[1])
        ...
        return self

Then, this random column would also be dropped in `predict` etc. Like you said, it assumes that the `fit` and `predict` datasets have the same shape. But it's a fair assumption imho as it is assumed everywhere in scikit-learn as well.

**Comment by soumyadsanyal** [comment]
Sounds great, thanks!
